File: leetcode/p0148/Solution.java
/**
https://leetcode.com/problems/sort-list/

Given the head of a linked list, return the list after sorting it in ascending order.

Follow up: Can you sort the linked list in O(n logn) time and O(1) memory (i.e. constant space)?

 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        List<ListNode> list = new ArrayList<>();
        
        ListNode cur = head;
        while(cur != null) {
            list.add(cur);
            cur = cur.next;
        }
        
        if (list.size() <= 1) {
            return head;
        }
        
        Collections.sort(list, (node0, node1) -> {
            return node0.val - node1.val;
        });
        
        ListNode sortedHead = list.get(0);
        for (int i = 0; i < list.size() - 1; ++i) {
            list.get(i).next = list.get(i + 1);
        }
        list.get(list.size() - 1).next = null;
        return sortedHead;
    }
}

File: leetcode/p0148/Solution.kt
class Solution {
    fun sortList(head: ListNode?): ListNode? {
        if (head == null || head.next == null) return head
        
        val mid = getMidAndSplit(head)
        val left = sortList(head)
        val right = sortList(mid)
        return mergeList(left, right)
    }
    private fun printList(head: ListNode?) {
        var cur = head
        while(cur != null) {
            print(cur.`val`)
            print("->")
            cur = cur.next
        }
        println()
    }
    private fun getMidAndSplit(head: ListNode?): ListNode? {
        if (head == null) return null
        
        //printList(head)
        var prev = head
        var slow = head
        var fast = head
        while (slow != null && fast != null && fast.next != null) {
            prev = slow
            slow = slow.next
            fast = fast.next.next
        }
        prev?.next = null
        return slow
    }
    
    private fun mergeList(left:ListNode?, right:ListNode?): ListNode? {
        val dummy = ListNode()
        var prev = dummy
        var l = left
        var r = right
        var cur:ListNode? = null
        while (l != null && r != null) {
            if (l.`val` < r.`val`) {
                cur = l
                l = l.next
            } else {
                cur = r
                r = r.next
            }
            prev.next = cur
            prev = cur
        }
        if (l != null) {
            prev.next = l
        } else if (r != null) {
            prev.next = r
        }
        return dummy.next
    }
}
